Added beginner friendly documentation

joybhallaa opened the pull request:
As I have elaborated in #4451 , I thought the `contributing.rst` is not easy to understand for beginners who are new to Git/GitHub and open source. I have added a section `Git & GitHub` inside the `Writing Patches` section. It will help contributors to setup Scrapy for contributing. 

I have run all tests using `tox`, tests have passed.

Fixes #4451

joybhallaa commented:
@Gallaecio thanks for the review, will work on the requested changes.

joybhallaa commented:
@Gallaecio ignore this commit, I mixed the work of 2 branches. 
I will fix this.

Gallaecio commented on docs/contributing.rst:
```diff
@@ -82,6 +82,43 @@ Writing patches
 
 The better a patch is written, the higher the chances that it'll get accepted and the sooner it will be merged.
 
+If you're new to scrapy or open source development, you would need to first learn what Git is and how Git & Github works.
```
```suggestion
If you're new to Scrapy or open source development, you would need to first learn what Git is and how Git & GitHub work.
```

Gallaecio commented on docs/contributing.rst:
```diff
@@ -82,6 +82,36 @@ Writing patches
+
+After installing Scrapy, you will have to create a branch in which you will do your work::
+
+    git checkout -b my-branch-name
```
Before branching, we should probably ask users to configure the upstream Scrapy repository as the `upstream` remote in their clone, and then recommend them to create branches the following way:

```bash
git fetch upstream
git checkout upstream/master -b new-branch-name
```

Otherwise, the second time they create a branch, it will probably be based on a previous, different set of changes to Scrapy that may not have been merged yet.

joybhallaa commented on docs/contributing.rst:
```diff
@@ -82,6 +82,36 @@ Writing patches
+
+#. Clone_ your fork
+
+#. Adding_ & committing changes 
```
I think the user should create a branch after setting up Scrapy on his machine.

Gallaecio replied:
Wouldn’t then the changes committed in this step be committed in the `master` branch of the contributor local clone?

joybhallaa replied:
My bad, I got confused. After the user has cloned the forked and cloned the repo, after that they should create a branch and work on that branch. 
Am I right ?

Gallaecio replied:
Yes, the point is not to “add and commit” before you have set up a branch. Setting up Scrapy may (and probably should) come before branch creation, but commits should definitely happen _after_ branch creation.

joybhallaa commented on docs/contributing.rst:
```diff
@@ -82,6 +82,36 @@ Writing patches
+
+After the above steps, you will need to install Scrapy in your machine by running::
+
+    pip install .
```
I agree, virtual environments are great for Python development !

joybhallaa commented:
@Gallaecio I've done a few things in this commit:
1. Added the `configure` step after the `clone` step
2. Moved the installation part to the steps
3. Added `branches` step at the end

Generated documentation using `tox -e docs`, and all tests have passed locally

Gallaecio commented on docs/contributing.rst:
```diff
@@ -80,6 +80,42 @@ guidelines when you're going to report a new bug.
+
+#. Clone_ your fork
+
+#. Configure_ upstream repository
```
```suggestion
#. Configure_ the upstream repository
```

Gallaecio commented on docs/contributing.rst:
```diff
@@ -80,6 +80,42 @@ guidelines when you're going to report a new bug.
+  
+         pip install -e .
+
+   NOTE: Consider setting up a `Virtual Environment`_ before installing Scrapy. Virtual environment is used to create an isolated environment for 
```
I wonder if, instead of suggesting this in a note, we should make the creation of a virtual environment a step of its own before installing Scrapy.

Gallaecio commented on docs/contributing.rst:
```diff
@@ -80,6 +80,42 @@ guidelines when you're going to report a new bug.
+   NOTE: Consider setting up a `Virtual Environment`_ before installing Scrapy. Virtual environment is used to create an isolated environment for 
+   python projects
+
+#. Then, create branches in the following way::
```
Maybe this step (and the next) should be out of the list of steps above, since these steps belong to a different process that must be followed every time that you start working on a new feature or bug fix.

joybhallaa replied:
@Gallaecio I could put it after the steps, to tell the user to install it?

Gallaecio replied:
Install what? I’m confused :confused:, my comment is about branch creation, not the installation of anything:

joybhallaa replied:
Okay, so I should put the branch creation after the steps. Is that correct ?

joybhallaa commented:
@Gallaecio if you're free, then let me know if the latest commit is okay. 
I am working on 2 more issues so I would like to close this ASAP :)

elacuesta commented on docs/contributing.rst:
```diff
@@ -80,6 +80,45 @@ guidelines when you're going to report a new bug.
+   NOTE: Consider setting up a `Virtual Environment`_ before installing Scrapy. Virtual environment is used to create an isolated environment for 
+   python projects
+
+.. _Clone: https://help.github.com/en/github/creating-cloning-and-archiving-repositories/cloning-a-repository
```
Should these definitions be at the end of the file, or at least, at the end of the section? For instance, `add_` is used after it's defined.

joybhallaa replied:
@elacuesta ,you're right. 
The definitions should be grouped together at the end of the section, as it will be easier to change definitions later, and it will look more organised.

elacuesta commented on docs/contributing.rst:
```diff
@@ -80,6 +80,45 @@ guidelines when you're going to report a new bug.
+
+When done with your changes, add_ and commit your changes.
+
+Before submitting your work, you should first run `tests`_ locally.
+
+Please make sure tests are passing before committing, this makes the reviewer's and maintainers work easy.
```
I'm not sure these lines are needed, I think they are included in the `* pass all unit-tests. See Running tests below` bullet point below ("Well-written patches should" bullet list).

joybhallaa replied:
okay, I'll remove them

joybhallaa commented:
`Warning, treated as error:
/docs/topics/settings.rst:1119:duplicate setting description of REDIRECT_MAX_TIMES, other instance in topics/downloader-middleware`

Getting this error when running **tox -e docs**, what should I do?

elacuesta commented:
> `Warning, treated as error: /docs/topics/settings.rst:1119:duplicate setting description of REDIRECT_MAX_TIMES, other instance in topics/downloader-middleware`

That was addressed in https://github.com/scrapy/scrapy/pull/4480#discussion_r406947675, merging `master` into your branch should solve it.

joybhallaa commented:
@elacuesta I've made some mistake, don't check the changes yet

elacuesta commented:
No need, I just rebased onto master and squashed all commits :+1:

joybhallaa commented:
@elacuesta check the latest commit when you're free
